### ISSUE TITLE
Inquiry on Code Implementation of "Context-Parallelism Distributed Inference" and "Logits-Masked Language Modeling Head"

### ISSUE THREAD

**Issue by xinyanghuang7** [VITA-MLLM/Long-VITA#2]
Thank you for your outstanding open-source contribution and the success of your project. 

In your technical report, you mentioned implementing "context-parallelism distributed inference" and a "logits-masked language modeling head" to scale up the infinite number of image and text tokens for model deployment. 

Could you please specify which parts of the codebase correspond to these two techniques?

Looking forward to your response.

**Comment by shenyunhang** [comment]
hi @xinyanghuang7 ,

Thank you for your attention.

Both designs are enabled in the Megatron inference script:
```
scripts/megatron/qwen25/inference_qwen25_14b_intern_300m_server_cp.sh
```


Here is a brief explain how they work in the codebase.


- context-parallelism distributed inference

This part is enabled by setting `context-parallel-size` to a value larger than `1` in the inference script.

During inference, we first segment the input tokens to different devices in [`get_batch_on_this_cp_rank`](https://github.com/VITA-MLLM/Long-VITA/blob/main/long_vita_megatron/inference/text_generation/generation.py#L137) function.

The model will use the ring attention method to compute self-attention results.

Then, after the model forward, we use the [`sync_output`](https://github.com/VITA-MLLM/Long-VITA/blob/main/long_vita_megatron/inference/text_generation/generation.py#L179) function to gather the predicted logit from different devices.

- logits-masked language modeling head

This part is enabled by using `--logit-mask` in the inference script.

Before model forward, we will generate a [`logit_mask`](https://github.com/VITA-MLLM/Long-VITA/blob/main/long_vita_megatron/inference/text_generation/generation.py#L141).

`logit_mask` indicates the position of logits should be computed.

During model forward, `logit_mask` is passed to [`output_layer`](https://github.com/VITA-MLLM/Long-VITA/blob/main/long_vita_megatron/core/models/multimodal/gpt_vl_model.py#L339), i.e., the language modeling head.

And this language modeling head is a modified `ColumnParallelLinear` layer, and its forward and backward functions are implemented at [here](https://github.com/VITA-MLLM/Long-VITA/blob/main/long_vita_megatron/core/tensor_parallel/layers.py#L825) and [here](https://github.com/VITA-MLLM/Long-VITA/blob/main/long_vita_megatron/core/tensor_parallel/layers.py#L416).

**Comment by xinyanghuang7** [comment]
Thank you for your detailed and timely response, which was very helpful to me and perfectly solved my problem.
